plutus-ledger-api/src/Plutus/V1/Ledger/Address.hs

```diff
@@ -11,50 +11,77 @@
 module Plutus.V1.Ledger.Address (
     Address (..),
     pubKeyAddress,
+    pubKeyHashAddress,
     scriptAddress,
     scriptHashAddress,
+    toPubKeyHash,
+    toValidatorHash,
+    stakingCredential
     ) where
 
-import           Codec.Serialise.Class     (Serialise)
-import           Control.DeepSeq           (NFData)
-import           Data.Aeson                (FromJSON, FromJSONKey (..), ToJSON, ToJSONKey (..))
-import           Data.Hashable             (Hashable)
+import           Codec.Serialise.Class       (Serialise)
+import           Control.DeepSeq             (NFData)
+import           Data.Aeson                  (FromJSON, FromJSONKey (..), ToJSON, ToJSONKey (..))
+import           Data.Hashable               (Hashable)
 import           Data.Text.Prettyprint.Doc
-import           GHC.Generics              (Generic)
-import qualified PlutusTx                  as PlutusTx
-import qualified PlutusTx.Eq               as PlutusTx
+import           GHC.Generics                (Generic)
+import qualified PlutusTx                    as PlutusTx
+import qualified PlutusTx.Bool               as PlutusTx
+import qualified PlutusTx.Eq                 as PlutusTx
 
+import           Plutus.V1.Ledger.Credential (Credential (..), StakingCredential)
 import           Plutus.V1.Ledger.Crypto
-import           Plutus.V1.Ledger.Orphans  ()
+import           Plutus.V1.Ledger.Orphans    ()
 import           Plutus.V1.Ledger.Scripts
 
--- | A payment address using a hash as the id.
-data Address = PubKeyAddress PubKeyHash
-    | ScriptAddress ValidatorHash
+-- | Address with two kinds of credentials, normal and staking
+data Address = Address{ addressCredential :: !Credential, addressStakingCredential :: !(Maybe StakingCredential)}
     deriving stock (Eq, Ord, Show, Generic)
     deriving anyclass (ToJSON, FromJSON, ToJSONKey, FromJSONKey, Serialise, Hashable, NFData)
 
 instance Pretty Address where
-    pretty (PubKeyAddress pkh) = "PubKeyAddress:" <+> pretty pkh
-    pretty (ScriptAddress vh)  = "ScriptAddress:" <+> pretty vh
+    pretty (Address cred stakingCred) = pretty cred <+> pretty stakingCred
```

Comment:
Should we do something prettier than this?

Reply:
I need to check what it actually looks like :D maybe I should add an "Address:" at the beginning.

Reply:
ok I made it a little nicer.